core/partitioning/partitioning.cpp

```diff
@@ -75,7 +75,8 @@ void find_all_fallback_nodes(std::unordered_map<torch::jit::Node*, int>& fallbac
     q.pop();
     // for every node that produces this fallback node's NonTensor input, they should fallback too
     for (auto input : cur_node->inputs()) {
-      if (!isTensor(input) && fallback_nodes.insert({input->node(), 4}).second) {
+      if (!isTensor(input) && input->node()->kind() != torch::jit::prim::Constant &&
```

Comment:
I think this input->node() traversal misses cases where an op modifies an input.

Ex.
%0 : ListConstruct()
%1 : aten::append(%0, %val)
%2 : aten::append(%1, %val)
%3 : aten::cat(%0)

Looking at just the input->node() of the cat skips over the appends which could result in a change in model behavior.

Reply:
Hi @mfeliz-cruise did you hit this kind of issue in your model?
I built a model locally with this graph:
```
graph(%x.1 : Tensor,
      %y.1 : Tensor):
  %2 : int = prim::Constant[value=0]() 
  %mod_list.1 : Tensor[] = prim::ListConstruct(%x.1)
  %z.1 : Tensor = aten::__getitem__(%mod_list.1, %2) 
  %5 : Tensor[] = aten::append(%mod_list.1, %x.1) 
  %6 : Tensor[] = aten::append(%mod_list.1, %y.1) 
  %7 : Tensor[] = aten::append(%mod_list.1, %z.1) 
  %8 : Tensor = aten::cat(%mod_list.1, %2) 
  return (%8)
```
with forced fallback on aten::cat, the model runs fine. It's true that the input->node() of aten::cat could skip over the aten::append, however, since aten::cat dependency node is prim::ListConstruct, and prim::ListConstruct's output is used later by aten::append, as a result, aten::append will also fallback.

I think there are several cases and these cases could all be covered: 

%0: nodeA
%1: nodeB_modifyinput(%0)
%2: nodeC(%1)

1: nodeC fallbacks, then there are 2 senarios: 
1.1: %1 is Tensor, in this case, nodeA should not fallback, nodeB which modifies input should also not fallback. We cannot get them fallback by traversing. 
1.2: %1 is not Tensor, which could be a TensorList like what happens here. So, in this case, nodeA will fallback. Since nodeA fallback, nodeB which modifies %0 will also fallback finally because nodeB is in nodeA's output list. 

2: nodeB fallbacks, 2 cases too:
2.1 if the input that nodeB is modifying is a Tensor, then nodeA and nodeC will not fallback.
2.2 if the input that nodeB is modifying is not a Tensor, then nodeA will fallback since nodeA is nodeB's input, since nodeA fallbacks, nodeC will also fallback since it's in nodeA's output. I also did a local test with forced fallback on aten::append, which also works fine. 

Did I miss any case here in the analysis?

Reply:
I think you're right that these cases should be handled by the forward fallback propagation from LIstConstruct. The cases where I've seen issues are caused by GetDependencyNodes where the ListConstruct node is identified as a dependency and copied into the segment of the cat without the appends, changing the behavior of the model.

Reply:
Thanks Bo, could we use the AliasInfo embedded in the node->schema to find these ops on the fly rather than looking it up from a static list? That seems like it could be more robust.

I'm looking at what's done in AliasDB here to get AliasInfo for an input: https://github.com/pytorch/pytorch/blob/d26c575ff581e4df0e9c72b339a25999c6cae59e/torch/csrc/jit/ir/alias_analysis.cpp#L816

And then check if it's a write:
https://github.com/pytorch/pytorch/blob/d26c575ff581e4df0e9c72b339a25999c6cae59e/torch/csrc/jit/ir/alias_analysis.cpp#L847

https://caffe2.ai/doxygen-c/html/classc10_1_1_alias_info.html

Reply:
This would also let you check which inputs are modified by an op rather than assuming if a value is used by a modifying op it will be modified.

Reply:
Thanks @mfeliz-cruise ! This is really helpful! 
The method your proposed is much more gentle than the current solution. 
Thanks for sharing this AliasInfo APIs, I'm going to take a look and try to use them to find the modifying nodes.